File: rust/fluentcards/src/builders/rich_text_block_builder.rs
use serde_json::Value;

use crate::enums::*;
use crate::models::Card;
use super::text_run_builder::TextRunBuilder;

/// Builds a RichTextBlock Adaptive Card element.
pub struct RichTextBlockBuilder {
    data: Card,
}

impl RichTextBlockBuilder {
    pub(crate) fn new() -> Self {
        let mut data = Card::new();
        data.insert("type".into(), Value::String("RichTextBlock".into()));
        data.insert("inlines".into(), Value::Array(Vec::new()));
        RichTextBlockBuilder { data }
    }

    pub fn with_id(&mut self, id: &str) -> &mut Self {
        self.data.insert("id".into(), Value::String(id.into()));
        self
    }

    pub fn with_horizontal_alignment(&mut self, alignment: HorizontalAlignment) -> &mut Self {
        self.data
            .insert("horizontalAlignment".into(), alignment.into());
        self
    }

    pub fn with_spacing(&mut self, spacing: Spacing) -> &mut Self {
        self.data.insert("spacing".into(), spacing.into());
        self
    }

    /// Adds a plain string inline.
    pub fn add_text(&mut self, text: &str) -> &mut Self {
        if let Some(Value::Array(inlines)) = self.data.get_mut("inlines") {
            inlines.push(Value::String(text.into()));
        }
        self
    }

    /// Adds a TextRun inline configured by the provided closure.
    pub fn add_text_run(
        &mut self,
        configure: impl FnOnce(&mut TextRunBuilder),
    ) -> &mut Self {
        let mut tb = TextRunBuilder::new();
        configure(&mut tb);
        if let Some(Value::Array(inlines)) = self.data.get_mut("inlines") {
            inlines.push(Value::Object(tb.build()));
        }
        self
    }

    pub fn build(&self) -> Card {
        self.data.clone()
    }
}

File: swift/Sources/FluentCards/RichTextBlockBuilder.swift
/// RichTextBlockBuilder builds a RichTextBlock Adaptive Card element.
public final class RichTextBlockBuilder {
    var data: [String: Any] = ["type": "RichTextBlock", "inlines": [Any]()]

    public init() {}

    @discardableResult
    public func withID(_ id: String) -> RichTextBlockBuilder {
        data["id"] = id
        return self
    }

    @discardableResult
    public func withHorizontalAlignment(_ alignment: HorizontalAlignment) -> RichTextBlockBuilder {
        data["horizontalAlignment"] = alignment.rawValue
        return self
    }

    @discardableResult
    public func withSpacing(_ spacing: Spacing) -> RichTextBlockBuilder {
        data["spacing"] = spacing.rawValue
        return self
    }

    /// Adds a plain string inline.
    @discardableResult
    public func addText(_ text: String) -> RichTextBlockBuilder {
        var inlines = data["inlines"] as! [Any]
        inlines.append(text)
        data["inlines"] = inlines
        return self
    }

    /// Adds a TextRun inline configured by the provided closure.
    @discardableResult
    public func addTextRun(_ configure: (TextRunBuilder) -> Void) -> RichTextBlockBuilder {
        let tb = TextRunBuilder()
        configure(tb)
        var inlines = data["inlines"] as! [Any]
        inlines.append(tb.build())
        data["inlines"] = inlines
        return self
    }

    public func build() -> Card {
        return data
    }
}
